### PR TITLE
Permissive mode support

### PR DESCRIPTION
**Pull request by ggriffiths**
Signed-off-by: Grant Griffiths <grant@portworx.com>

**What this PR does / why we need it**:
* Adds permissive mode core features

**In this PR is the following:**
* Permissive mode core
    * Create public user role
    * Replace empty token with public user info
    * Create public resources when public user
* Support for different modes
    * Disabled (no auth configured)
    * Permissive (auth configured)
    * Strict (via CLI/API or configuration)

**The following features will be implemented in separate PRs:**
* tooling for admin to see which vols are public
* be able to set volumes as public

### CODE REVIEW

File: api/server/volume_test.go [hunk 1]
```diff
@@ -2415,37 +2415,6 @@ func TestMiddlewareVolumeDeleteSuccess(t *testing.T) {
 
 }
 
-func TestMiddlewareVolumeDeleteFailure(t *testing.T) {
```
Comment (ggriffiths) [inline api/server/volume_test.go]: This test is no longer applicable. It was failing because a token was not provided.

File: api/api.go [hunk 2]
```diff
@@ -1098,6 +1098,11 @@ func (v *VolumeSpec) GetCloneCreatorOwnership(ctx context.Context) (*Ownership,
 
 	// If there is user information, then auth is enabled
 	if userinfo, ok := auth.NewUserInfoFromContext(ctx); ok {
+		// if a public user clones a volume, we should not update
```
Comment (lpabon) [inline api/api.go]: Remove this and let `OwnershipSetUsernameFromContext(ctx, nil)` handle returning the appropriate ownership object, which could be a public one or a `nil`

File: api/api.proto [hunk 3]
```diff
@@ -597,6 +597,10 @@ message Ownership {
     Admin = 2;
   }
 
+  message PublicAccessControl {
```
Comment (lpabon) [inline api/api.proto]: This is comments

File: api/ownership.go [hunk 4]
```diff
@@ -34,9 +34,14 @@ const (
 // a volume. It takes an ownership value if passed in by the user, then
 // sets the `owner` value to the user name referred to in the user context
 func OwnershipSetUsernameFromContext(ctx context.Context, srcOwnership *Ownership) *Ownership {
+
```
Comment (lpabon) [inline api/ownership.go]: delete this line

File: api/ownership.go [hunk 5]
```diff
@@ -34,9 +34,14 @@ const (
...
 	// Check if the context has information about the user. If not,
 	// then security is not enabled.
 	if userinfo, ok := auth.NewUserInfoFromContext(ctx); ok {
+		// Public users cannot provide ownership
+		if userinfo.Public {
+			return srcOwnership
```
Comment (lpabon) [inline api/ownership.go]: Return nil to be clear

File: api/ownership.go [hunk 6]
```diff
@@ -182,7 +187,8 @@ func (o *Ownership) HasAnOwner() bool {
 
 // IsPublic returns true if there is no ownership in this resource
 func (o *Ownership) IsPublic() bool {
-	return !o.HasAnOwner()
+	return (o.Acls != nil && o.Acls.Public != nil &&
```
Comment (lpabon) [inline api/ownership.go]: Simplify the code by adding it as a function `IsAccessPermittedByPublic(accessType)`

File: api/ownership.go [hunk 7]
```diff
@@ -182,7 +187,8 @@ func (o *Ownership) HasAnOwner() bool {
 
 // IsPublic returns true if there is no ownership in this resource
 func (o *Ownership) IsPublic() bool {
```
Comment (lpabon) [inline api/ownership.go]: Adjust it to support accessType

File: api/server/sdk/server_interceptors.go [hunk 8]
```diff
@@ -54,8 +54,22 @@ func (s *sdkGrpcServer) rwlockIntercepter(
 
 // Authenticate user and add authorization information back in the context
 func (s *sdkGrpcServer) auth(ctx context.Context) (context.Context, error) {
+	var token string
+	var err error
+
+	// public call attempted, add system.public user
+	if auth.IsPublic(ctx) {
+		return auth.ContextSaveUserInfo(ctx, &auth.UserInfo{
```
Comment (lpabon) [inline api/server/sdk/server_interceptors.go]: Create a function in auth to return the UserInfo for public with the appropriate RBAC.

Pass in the value of the PublicVolumeCreationDisabled bool into the new function and let it set the rules for the system.public role accordingly.

File: api/server/sdk/server_interceptors.go [hunk 9]
```diff
@@ -141,6 +155,12 @@ func (s *sdkGrpcServer) authorizationServerInterceptor(
 		"method":   info.FullMethod,
 	})
 
+	// If a public create call is attempted, but public volume creation is disabled, then deny
+	if auth.IsPublic(ctx) && s.config.Security.PublicVolumeCreationDisabled &&
```
Comment (lpabon) [inline api/server/sdk/server_interceptors.go]: We can remove this and let the RBAC system take care of it

File: cmd/osd/main.go [hunk 10]
```diff
@@ -483,9 +483,10 @@ func start(c *cli.Context) error {
 			Cluster:       cm,
 			StoragePolicy: sp,
 			Security: &sdk.SecurityConfig{
-				Role:           rm,
-				Tls:            tlsConfig,
-				Authenticators: authenticators,
+				Role:                         rm,
+				Tls:                          tlsConfig,
+				Authenticators:               authenticators,
+				PublicVolumeCreationDisabled: c.Bool("publicvolumecreationdisabled"),
```
Comment (lpabon) [inline cmd/osd/main.go]: `--public-volume-create-allowed=true` and set the default to `true`, then use `!c.Bool('...')`

File: pkg/auth/token.go [hunk 11]
```diff
@@ -109,3 +111,7 @@ func Token(
 
 	return signedtoken, nil
 }
+
+func IsPublic(ctx context.Context) bool {
+	return metautils.ExtractIncoming(ctx).Get("authorization") == ""
```
Comment (lpabon) [inline pkg/auth/token.go]: Use a constant for the string and investigate that the key is case-insensitive

File: pkg/role/sdkserviceapi.go [hunk 12]
```diff
@@ -91,6 +91,19 @@ var (
 				},
 			},
 		},
+
+		// system:public role is used for any unauthenticated user.
+		// They can only use standard volume lifecycle commands.
+		"system.public": {
```
Comment (lpabon) [inline pkg/role/sdkserviceapi.go]: Add support for `!` not allowed RBAC. `!*inspect*`. Remove `!` then send the string to `match(r[1:], s)`

```		"system.public": {
			&api.SdkRule{
				Services: []string{"volume"},
				Apis:     []string{"!create", "*"},
			},
			&api.SdkRule{
				Services: []string{"mountattach", "cloudbackup", "migrate"},
				Apis:     []string{"*"},
			},
			&api.SdkRule{
				Services: []string{"identity"},
				Apis:     []string{"version"},
			},
		},
```

Update the documentation in the Role API in the api.proto to support this, and show examples.

File: pkg/auth/token.go (hunk 11 above)
Comment (ggriffiths) [reply pkg/auth/token.go]: it does toLower: https://github.com/grpc-ecosystem/go-grpc-middleware/blob/master/util/metautils/single_key.go#L16

File: pkg/role/sdkserviceapi.go (hunk 12 above)
Comment (ggriffiths) [reply pkg/role/sdkserviceapi.go]: This will happen in a separate PR.

File: api/server/sdk/server_interceptors.go (hunk 8 above)
Comment (ggriffiths) [reply api/server/sdk/server_interceptors.go]: We decided to check PublicVolumeCreationDisabled on server startup and updating the system.public user RBAC accordingly.

File: api/api.go [hunk 13]
```diff
@@ -1131,7 +1130,7 @@ func (s *SdkStoragePolicy) IsPermitted(ctx context.Context, accessType Ownership
 }
 
 func (s *SdkStoragePolicy) IsPermittedFromUserInfo(user *auth.UserInfo, accessType Ownership_AccessType) bool {
-	if s.IsPublic() {
+	if s.IsPublic(Ownership_Read) {
```
Comment (ggriffiths) [inline api/api.go]: I chose to still pass in Ownership_Read for the Public checks, since Public is always Ownership_Read, right?

**Comment by ggriffiths** [comment]
Addressed feedback @lpabon , but have a question about the storage policy IsPublic function.

File: api/api.go (hunk 13 above)
Comment (ggriffiths) [reply api/api.go]: Updated all of these to pass in accessType instead.

File: api/ownership.go [hunk 14]
```diff
@@ -37,6 +37,10 @@ func OwnershipSetUsernameFromContext(ctx context.Context, srcOwnership *Ownershi
 	// Check if the context has information about the user. If not,
 	// then security is not enabled.
 	if userinfo, ok := auth.NewUserInfoFromContext(ctx); ok {
+		// Public users cannot provide ownership
+		if userinfo.Public {
```
Comment (lpabon) [inline api/ownership.go]: This probably should be an accessor func

File: api/server/sdk/credentials.go [hunk 15]
```diff
@@ -500,7 +500,7 @@ func (s CredentialServer) isPermitted(
 	}
 
 	// If ownership is missing then it is also public
-	if ownership == nil || ownership.IsPublic() {
+	if ownership == nil || ownership.IsPublic(api.Ownership_Read) {
```
Comment (lpabon) [inline api/server/sdk/credentials.go]: I think here we need to pass accessType

File: api/ownership.go (hunk 14 above)
Comment (ggriffiths) [reply api/ownership.go]: Added UserInfo.IsPublic()

File: api/server/sdk/credentials.go (hunk 15 above)
Comment (ggriffiths) [reply api/server/sdk/credentials.go]: Done

File: api/server/sdk/server_interceptors.go (hunk 9 above)
Comment (ggriffiths) [reply api/server/sdk/server_interceptors.go]: This will be refactored once we have support for negation in the RBAC system.

**Pull request merged**